edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -0,0 +1,423 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Channels;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Devices.Edge.Util;
+    using Microsoft.Extensions.Logging;
+    using uPLibrary.Networking.M2Mqtt;
+    using uPLibrary.Networking.M2Mqtt.Messages;
+
+    public class MqttBrokerConnector : IMqttBrokerConnector
+    {
+        const int ReconnectDelayMs = 2000;
+
+        readonly IComponentDiscovery components;
+        readonly ISystemComponentIdProvider systemComponentIdProvider;
+        readonly Dictionary<ushort, TaskCompletionSource<bool>> pendingAcks = new Dictionary<ushort, TaskCompletionSource<bool>>();
+
+        readonly object guard = new object();
+
+        Option<Channel<MqttPublishInfo>> publications;
+        Option<Task> forwardingLoop;
+        Option<MqttClient> mqttClient;
+
+        public MqttBrokerConnector(IComponentDiscovery components, ISystemComponentIdProvider systemComponentIdProvider)
+        {
+            this.components = Preconditions.CheckNotNull(components);
+            this.systemComponentIdProvider = Preconditions.CheckNotNull(systemComponentIdProvider);
+
+            // because of the circular dependency between MqttBridgeConnector and the producers,
+            // in this loop the producers get the IMqttBridgeConnector reference:
+            foreach (var producer in components.Producers)
```

Comment:
So, is the circular dependency because of IComponentDiscovery? Why can't the producer types take in a IMqttBrokerConnector? We don't seem to be using the producers anywhere else here.

Reply:
Some classes are producers and consumers in the same time. The MqttBrokerConnector needs the consumers (as it calls them with "handle()"), so the consumers need to be injected to the MqttBrokerConnector class (that happens trough the discovery class). These consumer classes, when they are also producers (e.g TwinHandler), need to get a reference to MqttBrokerConnector.

Simpler - TwinHandler needs a reference to MqttBrokerConnector as it needs to send messages out, but MqttBrokerConnector also needs a reference to TwinHandler, so it can call its "HandleAsync" method when a message comes in.

Reply:
If so, can we just get rid of the Discovery class and have a list of producers and list of consumers?

Reply:
that would not solve the circular dependency I think. When TwinHandler is instantiated, MqttBrokerConnection must already be an instance (so it can be injected into TwinHandler). But when MqttBrokerConnection is instantiated, it needs the consumer instances (which TwinHander is), so coming from this way, TwinHandler must be created first. Something must be created first, and the other thing needs an additional method call to obtain the first one.

Reply:
Hmm, but that seems to stem from the fact that we use the same class to implement IMessageProducer and IMessageConsumer.
Its a bit philosophical, but the SetConnector on the IMessageProducer doesn't seem necessary.. in fact feels like we don't need the IMessageProducer interface at all. 
But its okay for now, let's think about it later.

Reply:
The two interfaces can be merged, but not all producers are consumers and not all consumers are producers. Because of that, with a merged interface, we need to give "not implemented" implementations (either doing nothing or throwing an error). That is a typical sign of breaking the "interface segregation principle" from the solid principles - that is why am a bit against merging them. (I was crying deleting ISubscriber, but that at least was well aligned with IMessageConsumer) 

but even if it is merged, in the case of classes like TwinManager, I don't think that is solves the circular problem: the TwinHandler still needs an instance of the MqttBrokerConnector (because it uses it for sending messages), and the MqttBrokerConnector needs an instance of the TwinManager (because it wants to give incoming messages to it for processing) - so the circular reference is still there, making SetConnector a necessity.